diff --git a/dist/dev/index.cjs b/dist/dev/index.cjs
index 1ac04794f57a53e97262f101668452c272a8e6af..bdcf51346a50e4cb05151d8ab35e44e78ebf5a83 100644
--- a/dist/dev/index.cjs
+++ b/dist/dev/index.cjs
@@ -407,14 +407,7 @@ typeof URL.parse === 'function') {
       return ':';
     }
   };
-} else {
-  // fallback for IE11 support
-  let parsingNode = document.createElement('a');
-  protocolForUrl = url => {
-    parsingNode.href = url;
-    return parsingNode.protocol;
-  };
-}
+} 
 function sanitizeAttributeValue(element, attribute, value) {
   let tagName = null;
   if (value === null || value === undefined) {
diff --git a/dist/dev/index.js b/dist/dev/index.js
index 6d6c660960db5ab64bade5fd176c3339eb21df95..1eb1524fef522863aac2908d0b6d4f20f9bd2824 100644
--- a/dist/dev/index.js
+++ b/dist/dev/index.js
@@ -365,14 +365,7 @@ typeof URL.parse === 'function') {
       return ':';
     }
   };
-} else {
-  // fallback for IE11 support
-  let parsingNode = document.createElement('a');
-  protocolForUrl = url => {
-    parsingNode.href = url;
-    return parsingNode.protocol;
-  };
-}
+} 
 function sanitizeAttributeValue(element, attribute, value) {
   let tagName = null;
   if (value === null || value === undefined) {
diff --git a/index.ts b/index.ts
deleted file mode 100644
index 782461744bef5ea7acf6f5ff5ac22c425eff309a..0000000000000000000000000000000000000000
