diff --git a/dist/create-trace-benchmark.js b/dist/create-trace-benchmark.js
index 5918e8f7665b3e796ef88283fc40c2b3286a564f..e1a8768964de8b3c16d38bfb6280d836b45aba0b 100644
--- a/dist/create-trace-benchmark.js
+++ b/dist/create-trace-benchmark.js
@@ -45,9 +45,8 @@ function createTraceBenchmark(group, sampleTrace, options = {}) {
 }
 exports.default = createTraceBenchmark;
 function getCategories(isTrial, options) {
-    const categories = ['-*', ...defaultCategories];
+    const categories = [...defaultCategories];
     if (isTrial) {
-        categories.push(...captureAllDevtoolsTimelineCategories, ...captureCpuProfileCategories, captureCpuProfilesHiresCategory, captureFilmStripCategory, ...capturePaintProfileCategories);
         if (options.additionalTrialCategories) {
             categories.push(...options.additionalTrialCategories);
         }
