diff --git a/js/package.js b/js/package.js
index 91c757199e649b581e3064130f35f45d2af771a4..26d3e8bde91cd085e7a035a8520c8485e660be24 100644
--- a/js/package.js
+++ b/js/package.js
@@ -413,7 +413,8 @@ class Package {
         }
         let ensureModuleApiPolyfill = semver_1.default.satisfies(emberSource.pkg.version, '<3.27.0', { includePrerelease: true });
         let templateCompilerPath = emberSource.absolutePaths
-            .templateCompiler;
+            ? emberSource.absolutePaths.templateCompiler
+            : require.resolve('ember-source/ember-template-compiler/index.js');
         const babelPluginPrecompile = ensureModuleApiPolyfill
             ? [
                 require.resolve('babel-plugin-htmlbars-inline-precompile'),
