// glimmer-runtime/lib/compiled/opcodes/component.ts
import { Opcode, OpcodeJSON, UpdatingOpcode } from '../../opcodes';
import { Component, ComponentManager, ComponentDefinition } from '../../component/interfaces';
import { VM, UpdatingVM } from '../../vm';
import { CompiledArgs, EvaluatedArgs } from '../../compiled/expressions/args';
import { Templates } from '../../syntax/core';
import { layoutFor } from '../../compiler';
import { InternedString, dict } from 'glimmer-util';

export interface OpenComponentOptions {
  definition: ComponentDefinition<any>;
  args: CompiledArgs;
  shadow: InternedString[];
  templates: Templates;
}

export class OpenComponentOpcode extends Opcode {
  public type = "open-component";
  public definition: ComponentDefinition<any>;
  public args: CompiledArgs;
  public shadow: InternedString[];
  public templates: Templates;

  constructor({ definition, args, shadow, templates }: OpenComponentOptions) {
    super();
    this.definition = definition;
    this.args = args;
    this.shadow = shadow;
    this.templates = templates;
  }

  evaluate(vm: VM) {
    let { args: rawArgs, shadow, definition, templates } = this;
    let args = rawArgs.evaluate(vm);

    let manager = definition.manager;
    let component = manager.create(definition, args.named);
    let selfRef = vm.env.rootReferenceFor(manager.getSelf(component));

    let callerScope = vm.scope();

    // pass through the list of outer attributes to shadow from the
    // invocation site, as well as the component definition as internal
    // arguments.
    args.internal = args.internal || dict<any>();
    args.internal['shadow'] = shadow;
    args.internal['definition'] = definition;
    args.internal['component'] = component;

    let layout = layoutFor(definition, vm.env);

    vm.pushRootScope(selfRef, layout.symbols);
    vm.invokeLayout({ templates, args, shadow, layout, callerScope });
    vm.env.didCreate(component, manager);
    vm.updateWith(new UpdateComponentOpcode({ name: definition.name, component, manager, args }));
  }

  toJSON(): OpcodeJSON {
    return {
      guid: this._guid,
      type: this.type,
      args: [JSON.stringify(this.definition.name)]
    };
  }
}

export class UpdateComponentOpcode extends UpdatingOpcode {
  public type = "update-component";

  private name: string;
  private component: Component;
  private manager: ComponentManager<any>;
  private args: EvaluatedArgs;

  constructor({ name, component, manager, args } : { name: string, component: Component, manager: ComponentManager<any>, args: EvaluatedArgs }) {
    super();
    this.name = name;
    this.component = component;
    this.manager = manager;
    this.args = args;
  }

  evaluate(vm: UpdatingVM) {
    let { component, manager, args } = this;
    manager.update(component, args.named);
    vm.env.didUpdate(component, manager);
  }

  toJSON(): OpcodeJSON {
    return {
      guid: this._guid,
      type: this.type,
      args: [JSON.stringify(this.name)]
    };
  }
}

export class DidCreateElementOpcode extends Opcode {
  public type = "did-create-element";

  evaluate(vm: VM) {
    let args = vm.frame.getArgs();
    let internal = args.internal;
    let definition: ComponentDefinition<any> = internal['definition'];
    let manager = definition.manager;
    let component: Component = internal['component'];

    manager.didCreateElement(component, vm.stack().element);
  }

  toJSON(): OpcodeJSON {
    return {
      guid: this._guid,
      type: this.type,
      args: ["$ARGS"]
    };
  }
}

// Slow path for non-specialized component invocations. Uses an internal
// named lookup on the args.
export class ShadowAttributesOpcode extends Opcode {
  public type = "shadow-attributes";

  evaluate(vm: VM) {
    let args = vm.frame.getArgs();
    let internal = args.internal;
    let shadow: InternedString[] = internal['shadow'];
    // let definition: ComponentDefinition<any> = internal['definition'];

    let named = args.named;

    if (!shadow) return;

    shadow.forEach(name => {
      vm.stack().setAttribute(name, named.get(name));
    });
  }

  toJSON(): OpcodeJSON {
    return {
      guid: this._guid,
      type: this.type,
      args: ["$ARGS"]
    };
  }
}

export class CloseComponentOpcode extends Opcode {
  public type = "close-component";

  evaluate(vm: VM) {
    vm.popScope();
  }
}
