// glimmer-runtime/lib/compiled/opcodes/component.ts
import { Opcode } from '../../opcodes';
import { ComponentDefinition } from '../../component/interfaces';
import { UpdateAttributeOpcode } from './dom';
import { VM } from '../../vm';
import { CompiledArgs } from '../../compiled/expressions/args';
import { Templates } from '../../syntax/core';
import { InternedString, intern, dict } from 'glimmer-util';
import { CONST_REFERENCE } from 'glimmer-reference';

interface OpenComponentOptions {
  definition: ComponentDefinition;
  args: CompiledArgs;
  shadow: InternedString[];
  templates: Templates;
}

export class OpenComponentOpcode extends Opcode {
  public type = "open-component";
  public definition: ComponentDefinition;
  public args: CompiledArgs;
  public shadow: InternedString[];
  public templates: Templates;

  constructor({ definition, args, shadow, templates }: OpenComponentOptions) {
    super();
    this.definition = definition;
    this.args = args;
    this.shadow = shadow;
    this.templates = templates;
  }

  evaluate(vm: VM) {
    let { args: rawArgs, shadow, definition, templates } = this;
    let args = rawArgs.evaluate(vm);

    let callerScope = vm.scope();

    // pass through the list of outer attributes to shadow from the
    // invocation site, as well as the component definition as internal
    // arguments.
    args.internal = args.internal || dict<any>();
    args.internal['shadow'] = shadow;
    args.internal['definition'] = definition;

    vm.pushRootScope(definition.layout.symbolTable.size);
    vm.invokeLayout({ templates, args, shadow, definition, callerScope });
  }
}

// Slow path for non-specialized component invocations. Uses an internal
// named lookup on the args.
export class ShadowAttributesOpcode extends Opcode {
  public type = "shadow-attributes";

  evaluate(vm: VM) {
    let args = vm.frame.getArgs();
    let internal = args.internal;
    let shadow: InternedString[] = internal['shadow'];
    let definition: ComponentDefinition = internal['definition'];

    let named = args.named;

    definition.didCreateElement(vm);

    if (!shadow) return;

    shadow.forEach(name => {
      let reference = named.get(intern(`@${name}`));
      let value = reference.value();
      vm.stack().setAttribute(name, value);

      if (!reference[CONST_REFERENCE]) {
        vm.updateWith(new UpdateAttributeOpcode(vm.stack().element, name, reference, value));
      }
    });

  }
}

export class CloseComponentOpcode extends Opcode {
  public type = "close-component";

  evaluate(vm: VM) {
    vm.popScope();
  }
}
