// glimmer-runtime/lib/template.ts
import { UpdatableReference } from 'glimmer-reference';
import { SerializedTemplate } from 'glimmer-wire-format';
import { EntryPoint, Layout } from './compiled/blocks';
import { Environment } from './environment';
import { ElementStack } from './builder';
import VM from './vm/append';
import Scanner from './scanner';

interface TemplateOptions {
  raw: EntryPoint;
}

interface RenderOptions {
  hostOptions?: Object;
  appendTo: Element;
}

interface EvaluateOptions {
  nextSibling?: Node;
}

export default class Template {
  static fromSpec(spec: SerializedTemplate, env: Environment): Template {
    let scanner = new Scanner(spec, env);
    return new Template({
      raw: scanner.scanEntryPoint()
    });
  }

  static layoutFromSpec(spec: SerializedTemplate, env: Environment): Layout {
    let scanner = new Scanner(spec, env);
    return scanner.scanLayout();
  }

  raw: EntryPoint;
  meta: Object;

  constructor({ raw }: TemplateOptions) {
    this.raw = raw;
  }

  prettyPrint() {
  }

  render(self: any, env: Environment, options: RenderOptions, blockArguments: any[]=null) {
    let elementStack = new ElementStack({ dom: env.getDOM(), parentNode: options.appendTo, nextSibling: null });
    let vm = VM.initial(env, { self: new UpdatableReference(self), elementStack, size: this.raw.symbolTable.size });

    this.raw.compile(env);
    return vm.execute(this.raw.ops);
  }
}
