// glimmer-runtime/lib/compiled/expressions/positional-args.ts
import { NULL_REFERENCE } from '../../references';
import { CompiledExpression } from '../expressions';
import VM from '../../vm/append';
import { PathReference } from 'glimmer-reference';

export abstract class CompiledPositionalArgs {
  static create({ values }: { values: CompiledExpression<any>[] }): CompiledPositionalArgs {
    if (values.length) {
      return new CompiledNonEmptyPositionalArgs({ values });
    } else {
      return COMPILED_EMPTY_POSITIONAL_ARGS;
    }
  }

  public type: string;
  abstract evaluate(vm: VM): EvaluatedPositionalArgs;
  abstract toJSON(): string;
}

class CompiledNonEmptyPositionalArgs extends CompiledPositionalArgs {
  public type = "positional-args";
  public values: CompiledExpression<any>[];

  constructor({ values }: { values: CompiledExpression<any>[] }) {
    super();
    this.values = values;
  }

  evaluate(vm: VM): EvaluatedPositionalArgs {
    let { values } = this;
    let valueReferences = new Array<any>(values.length);

    for (let i = 0; i < values.length; i++) {
      valueReferences[i] = <PathReference<any>>values[i].evaluate(vm);
    }

    return EvaluatedPositionalArgs.create({ values: valueReferences });
  }

  toJSON(): string {
    return `[${this.values.map(value => value.toJSON()).join(", ")}]`;
  }
}

export const COMPILED_EMPTY_POSITIONAL_ARGS = new (class extends CompiledPositionalArgs {
  public type = "empty-positional-args";

  evaluate(vm): EvaluatedPositionalArgs {
    return EvaluatedPositionalArgs.empty();
  }

  toJSON(): string {
    return `<EMPTY>`;
  }
});

export abstract class EvaluatedPositionalArgs {
  static empty(): EvaluatedPositionalArgs {
    return EVALUATED_EMPTY_POSITIONAL_ARGS;
  }

  static create({ values }: { values: PathReference<any>[] }) {
    return new NonEmptyEvaluatedPositionalArgs({ values });
  }

  public type: string;
  public values: PathReference<any>[];

  forEach(callback: (value: PathReference<any>) => void) {
    let values = this.values;
    for (let i = 0; i < values.length; i++) {
      callback(values[i]);
    }
  }

  abstract at(index: number): PathReference<any>;
  abstract value(): any[];
}

class NonEmptyEvaluatedPositionalArgs extends EvaluatedPositionalArgs {
  public values: PathReference<any>[];

  constructor({ values }: { values: PathReference<any>[] }) {
    super();
    this.values = values;
  }

  at(index: number): PathReference<any> {
    return this.values[index];
  }

  value(): any[] {
    let ret = new Array(this.values.length);
    for (let i = 0; i < this.values.length; i++) {
      ret[i] = this.values[i].value();
    }

    return ret;
  }
}

export const EVALUATED_EMPTY_POSITIONAL_ARGS = new (class extends EvaluatedPositionalArgs {
  at(): PathReference<any> {
    return NULL_REFERENCE;
  }

  value(): any[] {
    return [];
  }
});
